pyo3-ffi/src/datetime.rs

```diff
@@ -434,62 +431,29 @@ pub struct PyDateTime_CAPI {
 // Python already shares this object between threads, so it's no more evil for us to do it too!
 unsafe impl Sync for PyDateTime_CAPI {}
 
-/// Safe wrapper around the Python datetime C-API global. Note that this object differs slightly
-/// from the equivalent C object: in C, this is implemented as a `static PyDateTime_CAPI *`. Here
-/// this is implemented as a wrapper which implements [`Deref`] to access a reference to a
-/// [`PyDateTime_CAPI`] object.
-///
-/// In the [`Deref`] implementation, if the underlying object has not yet been initialized, the GIL
-/// will automatically be acquired and [`PyDateTime_IMPORT()`] called.
-pub static PyDateTimeAPI: _PyDateTimeAPI_impl = _PyDateTimeAPI_impl {
-    inner: GILOnceCell::new(),
-};
+pub static PyDateTimeAPI: UnsafeGILCell<PyDateTime_CAPI> = UnsafeGILCell::new(ptr::null_mut());
```

Comment:
How about we make this static not `pub`, and then we have `pub fn PyDateTimeAPI() -> *mut PyDateTime_CAPI` to read the value of it?

The implementation itself can be as simple as

```rust
struct PyDateTimeAPISingleton(UnsafeCell<*mut PyDateTime_CAPI>);
unsafe impl Sync for PyDateTimeAPISingleton;

static PyDateTimeAPI_impl: PyDateTimeAPISingleton = PyDateTimeAPISingleton(UnsafeCell::new());

#[inline]
pub unsafe fn PyDateTimeAPI() -> *mut PyDateTime_CAPI {
    *PyDateTimeAPISingleton().get()
}
```

... and then `PyDateTime_IMPORT` in this file is the only place where this is ever written.

Reply:
What about providing a macro to do this? There's quite a bit of code around that uses `static mut` to store pointers to Python objects. E.g. https://github.com/ijl/orjson/blob/master/src/typeref.rs#L24 

(this is probably out of scope for this PR, but perhaps a nice future addition?)

Reply:
Something similar to orjson is exactly my kind of usecase :) A macro would be nice but I agree its probably better for a followup PR

Reply:
>  f I squint a bit I also think we could have static PyDateTimeAPI: PyDateTimeApiSingleton and PyDateTimeApiSingleton has a Deref impl which produces &*mut PyDateTime_CAPI. That might feel more like the Python FFI, as PyDateTimeAPI is then a static rather than a function?!

I opted not to do this because reading from the UnsafeCell is always unsafe (because even for reading the caller has to guarantee that the GIL is held). Instead I opted for your former suggestion